### PR TITLE
Make Restart Controller Generic

### PR DESCRIPTION
**Pull request by rohitsakala**
[#172404604](https://www.pivotaltracker.com/n/projects/2192232/stories/172404604)

### CODE REVIEW

File: pkg/kube/controllers/quarksrestart/restart_controller.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+				ctxlog.Errorf(ctx, "Failed to calculate reconciles for secret '%s/%s': %v", secret.Namespace, secret.Name, err)
+			}
+
+			for _, reconciliation := range reconciles {
+				ctxlog.NewMappingEvent(a.Object).Debug(ctx, reconciliation, "RestartController", a.Meta.GetName(), "secret")
+			}
+
+			return reconciles
+		}),
+	}, p)
```
Comment (manno) [inline pkg/kube/controllers/quarksrestart/restart_controller.go:86]: This watch needs to use a 'namespace predicate' now (`monitorednamespace.NewNSPredicate`).

Comment (rohitsakala) [reply pkg/kube/controllers/quarksrestart/restart_controller.go:86]: Done.

File: docs/controllers/quarks_restart.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+## Description
+
+The QuarksRestart controller is responsible for restarting kubernetes resources such as `StatefulSet` and `Deployment`. They are restarted whenever a secret referenced by these resources gets updated. 
+
+This feature also enables updating entangled pods whenever the link secrets get updated.
+
+#### Watches in Quarks Restart Controller
+
+- `Secret`: Updates which have the annotation `quarks.cloudfoundry.org/update-referenced-owner`
```
Comment (viovanov) [inline docs/controllers/quarks_restart.md:17]: Updates for secrets that have the annotation

File: docs/controllers/quarks_restart.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+#### Reconciliation in Quarks Restart Controller
+
+- adds restart annotation `quarks.cloudfoundry.org/restart-by-quarks` to `StatefulSet` or `Deployment` as appropriate.
```
Comment (viovanov) [inline docs/controllers/quarks_restart.md:21]: Should be `quarks.cloudfoundry.org/restart`

File: docs/entanglements.md [hunk 4]
```diff
@@ -83,6 +83,10 @@ apiVersion: v1
 kind: Secret
 metadata:
   name: link-test-nats-nats
+  annotations:
+    quarks.cloudfoundry.org/update-referenced-owner: "true"
```
Comment (viovanov) [inline docs/entanglements.md:87]: I don't think the name of the label makes sense - "update referenced owner".
We're not updating an "owner". Please rename to `quarks.cloudfoundry.org/restart-on-update`.

File: docs/controllers/quarks_restart.md (hunk 2 above)
Comment (rohitsakala) [reply docs/controllers/quarks_restart.md:17]: Done.

File: docs/controllers/quarks_restart.md (hunk 3 above)
Comment (rohitsakala) [reply docs/controllers/quarks_restart.md:21]: Done.

File: docs/entanglements.md (hunk 4 above)
Comment (rohitsakala) [reply docs/entanglements.md:87]: Done.

**Pull request merged**